File: cheat/src/ui/theme.rs
use std::fmt::Display;

use egui::{
    Color32, CornerRadius, Mesh, Painter, Pos2, Rect, Shadow, Stroke, Style, TextStyle, Vec2,
};
use serde::{Deserialize, Serialize};
use strum::EnumIter;

const fn rgb(r: u8, g: u8, b: u8) -> Color32 {
    Color32::from_rgb(r, g, b)
}

/// Every structural color of the gui window. Accent colors from [`crate::ui::color::Colors`]
/// stay independent of this, they are picked by the user.
#[derive(Clone, Copy)]
pub struct Palette {
    pub dark: bool,
    /// deepest background, used for text edits and scroll troughs
    pub backdrop: Color32,
    /// default panel and window background
    pub base: Color32,
    /// idle widget background
    pub surface: Color32,
    /// hovered widget background
    pub overlay: Color32,
    /// secondary text: hints, units, inactive labels
    pub subtext: Color32,
    /// primary text
    pub text: Color32,
    /// accent the theme ships with, applied when the theme is picked
    pub accent: Color32,
    /// default gradient stops of the theme
    pub gradient: [Color32; 2],
}

#[derive(Debug, Clone, Copy, Default, PartialEq, EnumIter, Serialize, Deserialize)]
pub enum Theme {
    #[default]
    Midnight,
    Mocha,
    TokyoNight,
    Nord,
    Gruvbox,
    RosePine,
    Dracula,
    Daylight,
    Custom,
}

impl Display for Theme {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(match self {
            Self::Midnight => "Midnight",
            Self::Mocha => "Mocha",
            Self::TokyoNight => "Tokyo Night",
            Self::Nord => "Nord",
            Self::Gruvbox => "Gruvbox",
            Self::RosePine => "Rosé Pine",
            Self::Dracula => "Dracula",
            Self::Daylight => "Daylight",
            Self::Custom => "Custom",
        })
    }
}

impl Theme {
    /// Palette of a built in theme. [`Theme::Custom`] has no palette of its own, it falls back
    /// to the default one as a starting point, see [`ThemeConfig::palette`].
    pub fn palette(&self) -> Palette {
        match self {
            Self::Midnight | Self::Custom => Palette {
                dark: true,
                backdrop: rgb(15, 15, 19),
                base: rgb(26, 26, 32),
                surface: rgb(46, 46, 58),
                overlay: rgb(72, 72, 90),
                subtext: rgb(150, 152, 165),
                text: rgb(246, 247, 251),
                accent: rgb(100, 150, 240),
                gradient: [rgb(18, 18, 26), rgb(42, 36, 64)],
            },
            Self::Mocha => Palette {
                dark: true,
                backdrop: rgb(17, 17, 27),
                base: rgb(24, 24, 37),
                surface: rgb(49, 50, 68),
                overlay: rgb(76, 78, 102),
                subtext: rgb(166, 173, 200),
                text: rgb(223, 230, 252),
                accent: rgb(203, 166, 247),
                gradient: [rgb(17, 17, 27), rgb(54, 42, 78)],
            },
            Self::TokyoNight => Palette {
                dark: true,
                backdrop: rgb(19, 20, 28),
                base: rgb(26, 27, 38),
                surface: rgb(41, 46, 66),
                overlay: rgb(67, 75, 108),
                subtext: rgb(154, 165, 208),
                text: rgb(208, 216, 252),
                accent: rgb(122, 162, 247),
                gradient: [rgb(19, 20, 28), rgb(36, 42, 82)],
            },
            Self::Nord => Palette {
                dark: true,
                backdrop: rgb(36, 41, 51),
                base: rgb(46, 52, 64),
                surface: rgb(67, 76, 94),
                overlay: rgb(90, 101, 124),
                subtext: rgb(192, 201, 216),
                text: rgb(240, 243, 249),
                accent: rgb(136, 192, 208),
                gradient: [rgb(36, 41, 51), rgb(62, 78, 102)],
            },
            Self::Gruvbox => Palette {
                dark: true,
                backdrop: rgb(22, 24, 25),
                base: rgb(33, 35, 35),
                surface: rgb(60, 56, 54),
                overlay: rgb(92, 84, 76),
                subtext: rgb(176, 161, 138),
                text: rgb(248, 234, 200),
                accent: rgb(250, 189, 47),
                gradient: [rgb(22, 24, 25), rgb(68, 54, 40)],
            },
            Self::RosePine => Palette {
                dark: true,
                backdrop: rgb(18, 16, 28),
                base: rgb(25, 23, 36),
                surface: rgb(40, 37, 60),
                overlay: rgb(70, 66, 92),
                subtext: rgb(152, 148, 180),
                text: rgb(233, 231, 250),
                accent: rgb(196, 167, 231),
                gradient: [rgb(18, 16, 28), rgb(62, 42, 70)],
            },
            Self::Dracula => Palette {
                dark: true,
                backdrop: rgb(25, 26, 34),
                base: rgb(33, 34, 44),
                surface: rgb(58, 61, 80),
                overlay: rgb(92, 96, 120),
                subtext: rgb(158, 162, 188),
                text: rgb(250, 250, 246),
                accent: rgb(189, 147, 249),
                gradient: [rgb(25, 26, 34), rgb(68, 50, 94)],
            },
            Self::Daylight => Palette {
                dark: false,
                backdrop: rgb(220, 224, 233),
                base: rgb(250, 251, 253),
                surface: rgb(230, 234, 242),
                overlay: rgb(202, 209, 223),
                subtext: rgb(94, 101, 119),
                text: rgb(16, 18, 26),
                accent: rgb(45, 98, 214),
                gradient: [rgb(238, 241, 248), rgb(206, 218, 242)],
            },
        }
    }
}

/// User editable palette behind [`Theme::Custom`]. Kept as plain colors so it round trips
/// through the config file unchanged.
#[derive(Clone, Serialize, Deserialize)]
#[serde(default)]
pub struct CustomPalette {
    pub dark: bool,
    pub backdrop: Color32,
    pub base: Color32,
    pub surface: Color32,
    pub overlay: Color32,
    pub subtext: Color32,
    pub text: Color32,
}

impl Default for CustomPalette {
    fn default() -> Self {
        Self::from_palette(&Theme::Midnight.palette())
    }
}

impl CustomPalette {
    pub fn from_palette(palette: &Palette) -> Self {
        Self {
            dark: palette.dark,
            backdrop: palette.backdrop,
            base: palette.base,
            surface: palette.surface,
            overlay: palette.overlay,
            subtext: palette.subtext,
            text: palette.text,
        }
    }

    fn to_palette(&self, accent: Color32, gradient: [Color32; 2]) -> Palette {
        Palette {
            dark: self.dark,
            backdrop: self.backdrop,
            base: self.base,
            surface: self.surface,
            overlay: self.overlay,
            subtext: self.subtext,
            text: self.text,
            accent,
            gradient,
        }
    }
}

#[derive(Debug, Clone, Copy, Default, PartialEq, EnumIter, Serialize, Deserialize)]
pub enum GradientDirection {
    #[default]
    Vertical,
    Horizontal,
    DiagonalDown,
    DiagonalUp,
}

impl Display for GradientDirection {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(match self {
            Self::Vertical => "Vertical",
            Self::Horizontal => "Horizontal",
            Self::DiagonalDown => "Diagonal Down",
            Self::DiagonalUp => "Diagonal Up",
        })
    }
}

impl GradientDirection {
    /// unit vector the gradient runs along, in screen space (y points down)
    fn axis(&self) -> Vec2 {
        const DIAGONAL: f32 = std::f32::consts::FRAC_1_SQRT_2;
        match self {
            Self::Vertical => Vec2::new(0.0, 1.0),
            Self::Horizontal => Vec2::new(1.0, 0.0),
            Self::DiagonalDown => Vec2::new(DIAGONAL, DIAGONAL),
            Self::DiagonalUp => Vec2::new(DIAGONAL, -DIAGONAL),
        }
    }
}

/// Type sizes in points, before [`ThemeConfig::text_scale`] is applied. The spread between
/// them is what gives the window a reading order instead of one flat wall of text.
const HEADING_SIZE: f32 = 21.0;
const BODY_SIZE: f32 = 15.0;
const BUTTON_SIZE: f32 = 15.0;
const SMALL_SIZE: f32 = 12.0;
const MONOSPACE_SIZE: f32 = 14.0;

#[derive(Clone, Serialize, Deserialize)]
#[serde(default)]
pub struct ThemeConfig {
    pub theme: Theme,
    pub custom: CustomPalette,
    pub gradient: bool,
    pub gradient_start: Color32,
    pub gradient_end: Color32,
    pub gradient_direction: GradientDirection,
    /// how opaque the panels sit on top of the gradient, 0..=1
    pub panel_opacity: f32,
    /// multiplier on every text size, 1.0 is the designed scale
    pub text_scale: f32,
    /// pulls secondary text towards the primary text color, 0..=1
    pub text_contrast: f32,
    pub corner_radius: u8,
    pub shadows: bool,
}

impl Default for ThemeConfig {
    fn default() -> Self {
        let palette = Theme::default().palette();
        Self {
            theme: Theme::default(),
            custom: CustomPalette::from_palette(&palette),
            gradient: true,
            gradient_start: palette.gradient[0],
            gradient_end: palette.gradient[1],
            gradient_direction: GradientDirection::DiagonalDown,
            panel_opacity: 0.72,
            text_scale: 1.0,
            text_contrast: 0.0,
            corner_radius: 6,
            shadows: true,
        }
    }
}

impl ThemeConfig {
    /// Resolved palette: the preset, or the user's own colors when [`Theme::Custom`] is picked.
    /// The accent stays owned by `Config::accent_color`, so it is not part of this.
    pub fn palette(&self) -> Palette {
        let preset = self.theme.palette();
        let gradient = [self.gradient_start, self.gradient_end];
        match self.theme {
            Theme::Custom => self.custom.to_palette(preset.accent, gradient),
            _ => Palette { gradient, ..preset },
        }
    }

    /// pull gradient stops from the current theme, used when the theme changes
    pub fn sync_gradient(&mut self) {
        let palette = self.theme.palette();
        self.gradient_start = palette.gradient[0];
        self.gradient_end = palette.gradient[1];
    }

    /// seed the custom palette from a built in theme, so editing starts somewhere sane
    pub fn seed_custom(&mut self, from: Theme) {
        self.custom = CustomPalette::from_palette(&from.palette());
    }

    /// panel background, translucent while a gradient is showing through
    pub fn panel_fill(&self, base: Color32) -> Color32 {
        if !self.gradient {
            return base;
        }
        let alpha = (self.panel_opacity.clamp(0.0, 1.0) * 255.0).round() as u8;
        Color32::from_rgba_unmultiplied(base.r(), base.g(), base.b(), alpha)
    }

    /// secondary text color, pulled towards the primary one by `text_contrast`
    fn secondary_text(&self, palette: &Palette) -> Color32 {
        lerp_color(
            palette.subtext,
            palette.text,
            self.text_contrast.clamp(0.0, 1.0),
        )
    }
}

pub fn lerp_color(from: Color32, to: Color32, t: f32) -> Color32 {
    let t = t.clamp(0.0, 1.0);
    let lerp = |a: u8, b: u8| (a as f32 + (b as f32 - a as f32) * t).round() as u8;
    Color32::from_rgb(
        lerp(from.r(), to.r()),
        lerp(from.g(), to.g()),
        lerp(from.b(), to.b()),
    )
}

/// Color for section headers: the primary text color tinted towards the accent, so headers
/// read as their own level instead of blending into the body text.
pub fn heading_color(style: &Style) -> Color32 {
    let text = style.visuals.widgets.noninteractive.fg_stroke.color;
    let accent = style.visuals.selection.bg_fill;
    lerp_color(text, accent, 0.55)
}

/// Font size for section headers, one step above body text.
pub fn heading_size(style: &Style) -> f32 {
    style
        .text_styles
        .get(&TextStyle::Body)
        .map_or(BODY_SIZE, |font| font.size)
        + 2.0
}

/// Black or white, whichever stays readable on top of `color`. Used for text drawn directly
/// on the accent fill, where a bright accent would otherwise get bright text on it.
fn contrasting_text(color: Color32, palette: &Palette) -> Color32 {
    let luminance =
        0.2126 * color.r() as f32 + 0.7152 * color.g() as f32 + 0.0722 * color.b() as f32;
    if luminance > 140.0 {
        palette.backdrop
    } else {
        palette.text
    }
}

/// Paints the window background gradient as a single quad. The corner colors are the
/// gradient sampled at each corner's projection onto the gradient axis, so the linear
/// interpolation the gpu does between them reproduces the gradient exactly.
pub fn paint_gradient(painter: &Painter, rect: Rect, config: &ThemeConfig) {
    if !config.gradient || rect.width() <= 0.0 || rect.height() <= 0.0 {
        return;
    }

    let axis = config.gradient_direction.axis();
    let corners = [
        rect.left_top(),
        rect.right_top(),
        rect.right_bottom(),
        rect.left_bottom(),
    ];
    let projections = corners.map(|corner: Pos2| corner.to_vec2().dot(axis));
    let min = projections.iter().copied().fold(f32::INFINITY, f32::min);
    let max = projections
        .iter()
        .copied()
        .fold(f32::NEG_INFINITY, f32::max);
    let span = max - min;

    let mut mesh = Mesh::default();
    for (corner, projection) in corners.iter().zip(projections) {
        let t = if span > f32::EPSILON {
            (projection - min) / span
        } else {
            0.0
        };
        mesh.colored_vertex(
            *corner,
            lerp_color(config.gradient_start, config.gradient_end, t),
        );
    }
    mesh.add_triangle(0, 1, 2);
    mesh.add_triangle(0, 2, 3);

    painter.add(mesh);
}

/// Applies a theme to an egui style. Replaces the previous hardcoded dark palette.
pub fn apply(style: &mut Style, config: &ThemeConfig, accent: Color32) {
    let palette = config.palette();
    let radius = CornerRadius::same(config.corner_radius);
    let secondary = config.secondary_text(&palette);

    style.interaction.selectable_labels = false;

    let scale = config.text_scale.clamp(0.7, 2.0);
    for (text_style, font) in style.text_styles.iter_mut() {
        font.size = scale
            * match text_style {
                TextStyle::Heading => HEADING_SIZE,
                TextStyle::Body => BODY_SIZE,
                TextStyle::Button => BUTTON_SIZE,
                TextStyle::Small => SMALL_SIZE,
                TextStyle::Monospace => MONOSPACE_SIZE,
                TextStyle::Name(_) => BODY_SIZE,
            };
    }

    style.spacing.item_spacing = Vec2::new(8.0, 6.0);
    style.spacing.button_padding = Vec2::new(10.0, 4.0);
    style.spacing.menu_margin = egui::Margin::same(6);
    style.spacing.indent = 18.0;

    style.visuals.dark_mode = palette.dark;
    // panels let the gradient through, floating windows and dropdowns stay opaque so
    // their text keeps its contrast
    style.visuals.window_fill = palette.base;
    style.visuals.panel_fill = config.panel_fill(palette.base);
    style.visuals.extreme_bg_color = palette.backdrop;
    style.visuals.faint_bg_color = palette.surface;
    style.visuals.code_bg_color = palette.backdrop;
    style.visuals.hyperlink_color = accent;
    style.visuals.weak_text_color = Some(secondary);
    style.visuals.window_corner_radius = radius;
    style.visuals.menu_corner_radius = radius;
    style.visuals.window_stroke = Stroke::new(1.0, palette.overlay);

    style.visuals.window_shadow = if config.shadows {
        Shadow {
            offset: [0, 6],
            blur: 18,
            spread: 0,
            color: Color32::from_black_alpha(96),
        }
    } else {
        Shadow::NONE
    };
    style.visuals.popup_shadow = style.visuals.window_shadow;

    let text_stroke = Stroke::new(1.0, palette.text);
    let secondary_stroke = Stroke::new(1.0, secondary);
    let border_stroke = Stroke::new(1.0, palette.overlay);

    style.visuals.selection.bg_fill = accent;
    style.visuals.selection.stroke = Stroke::new(1.0, contrasting_text(accent, &palette));

    let widgets = &mut style.visuals.widgets;

    // labels, separators and other non widget text: the brightest text in the window
    widgets.noninteractive.bg_fill = palette.surface;
    widgets.noninteractive.weak_bg_fill = palette.surface;
    widgets.noninteractive.bg_stroke = border_stroke;
    widgets.noninteractive.fg_stroke = text_stroke;
    widgets.noninteractive.corner_radius = radius;

    // idle controls sit one step back from labels
    widgets.inactive.bg_fill = palette.surface;
    widgets.inactive.weak_bg_fill = palette.surface;
    widgets.inactive.bg_stroke = border_stroke;
    widgets.inactive.fg_stroke = secondary_stroke;
    widgets.inactive.corner_radius = radius;

    widgets.hovered.bg_fill = palette.overlay;
    widgets.hovered.weak_bg_fill = palette.overlay;
    widgets.hovered.bg_stroke = Stroke::new(1.0, lerp_color(palette.overlay, accent, 0.5));
    widgets.hovered.fg_stroke = text_stroke;
    widgets.hovered.corner_radius = radius;
    widgets.hovered.expansion = 1.0;

    widgets.active.bg_fill = lerp_color(palette.overlay, accent, 0.45);
    widgets.active.weak_bg_fill = lerp_color(palette.overlay, accent, 0.45);
    widgets.active.bg_stroke = Stroke::new(1.0, accent);
    widgets.active.fg_stroke = text_stroke;
    widgets.active.corner_radius = radius;
    widgets.active.expansion = 1.0;

    widgets.open.bg_fill = palette.overlay;
    widgets.open.weak_bg_fill = palette.overlay;
    widgets.open.bg_stroke = border_stroke;
    widgets.open.fg_stroke = text_stroke;
    widgets.open.corner_radius = radius;
}

/// Reapplies a theme to a live context, for when the user changes it at runtime.
pub fn apply_to_ctx(ctx: &egui::Context, config: &ThemeConfig, accent: Color32) {
    ctx.all_styles_mut(|style| apply(style, config, accent));
}

File: native/src/ui/theme.cpp
#include "ui/theme.hpp"

#include <algorithm>
#include <cmath>

namespace dl::ui {
namespace {

constexpr Color rgb(std::uint8_t r, std::uint8_t g, std::uint8_t b) { return Color(r, g, b); }

/// black or white, whichever stays readable on top of `color`. used for text drawn directly
/// on the accent fill, where a bright accent would otherwise get bright text on it
Color contrasting_text(Color color, const Palette& palette) {
    const float luminance =
        0.2126f * static_cast<float>(color.r) + 0.7152f * static_cast<float>(color.g) +
        0.0722f * static_cast<float>(color.b);
    return luminance > 140.0f ? palette.backdrop : palette.text;
}

void set(ImGuiStyle& style, ImGuiCol slot, Color color) { style.Colors[slot] = color.vec4(); }

}  // namespace

Color lerp(Color from, Color to, float t) {
    t = std::clamp(t, 0.0f, 1.0f);
    const auto mix = [t](std::uint8_t a, std::uint8_t b) {
        const float value =
            static_cast<float>(a) + (static_cast<float>(b) - static_cast<float>(a)) * t;
        return static_cast<std::uint8_t>(std::lround(value));
    };
    return Color(mix(from.r, to.r), mix(from.g, to.g), mix(from.b, to.b), mix(from.a, to.a));
}

std::string_view theme_name(Theme theme) {
    switch (theme) {
        case Theme::Midnight: return "Midnight";
        case Theme::Mocha: return "Mocha";
        case Theme::TokyoNight: return "Tokyo Night";
        case Theme::Nord: return "Nord";
        case Theme::Gruvbox: return "Gruvbox";
        case Theme::RosePine: return "Rose Pine";
        case Theme::Dracula: return "Dracula";
        case Theme::Daylight: return "Daylight";
        case Theme::Custom: return "Custom";
    }
    return "Midnight";
}

Palette theme_palette(Theme theme) {
    switch (theme) {
        case Theme::Midnight:
        case Theme::Custom:
            return Palette{true,          rgb(15, 15, 19),    rgb(26, 26, 32),
                           rgb(46, 46, 58), rgb(72, 72, 90),  rgb(150, 152, 165),
                           rgb(246, 247, 251), rgb(100, 150, 240),
                           {rgb(18, 18, 26), rgb(42, 36, 64)}};
        case Theme::Mocha:
            return Palette{true,          rgb(17, 17, 27),    rgb(24, 24, 37),
                           rgb(49, 50, 68), rgb(76, 78, 102), rgb(166, 173, 200),
                           rgb(223, 230, 252), rgb(203, 166, 247),
                           {rgb(17, 17, 27), rgb(54, 42, 78)}};
        case Theme::TokyoNight:
            return Palette{true,          rgb(19, 20, 28),     rgb(26, 27, 38),
                           rgb(41, 46, 66), rgb(67, 75, 108),  rgb(154, 165, 208),
                           rgb(208, 216, 252), rgb(122, 162, 247),
                           {rgb(19, 20, 28), rgb(36, 42, 82)}};
        case Theme::Nord:
            return Palette{true,          rgb(36, 41, 51),     rgb(46, 52, 64),
                           rgb(67, 76, 94), rgb(90, 101, 124), rgb(192, 201, 216),
                           rgb(240, 243, 249), rgb(136, 192, 208),
                           {rgb(36, 41, 51), rgb(62, 78, 102)}};
        case Theme::Gruvbox:
            return Palette{true,          rgb(22, 24, 25),    rgb(33, 35, 35),
                           rgb(60, 56, 54), rgb(92, 84, 76),  rgb(176, 161, 138),
                           rgb(248, 234, 200), rgb(250, 189, 47),
                           {rgb(22, 24, 25), rgb(68, 54, 40)}};
        case Theme::RosePine:
            return Palette{true,          rgb(18, 16, 28),    rgb(25, 23, 36),
                           rgb(40, 37, 60), rgb(70, 66, 92),  rgb(152, 148, 180),
                           rgb(233, 231, 250), rgb(196, 167, 231),
                           {rgb(18, 16, 28), rgb(62, 42, 70)}};
        case Theme::Dracula:
            return Palette{true,          rgb(25, 26, 34),    rgb(33, 34, 44),
                           rgb(58, 61, 80), rgb(92, 96, 120), rgb(158, 162, 188),
                           rgb(250, 250, 246), rgb(189, 147, 249),
                           {rgb(25, 26, 34), rgb(68, 50, 94)}};
        case Theme::Daylight:
            return Palette{false,              rgb(220, 224, 233), rgb(250, 251, 253),
                           rgb(230, 234, 242), rgb(202, 209, 223), rgb(94, 101, 119),
                           rgb(16, 18, 26),    rgb(45, 98, 214),
                           {rgb(238, 241, 248), rgb(206, 218, 242)}};
    }
    return theme_palette(Theme::Midnight);
}

std::string_view gradient_direction_name(GradientDirection direction) {
    switch (direction) {
        case GradientDirection::Vertical: return "Vertical";
        case GradientDirection::Horizontal: return "Horizontal";
        case GradientDirection::DiagonalDown: return "Diagonal Down";
        case GradientDirection::DiagonalUp: return "Diagonal Up";
    }
    return "Vertical";
}

CustomPalette CustomPalette::from_palette(const Palette& palette) {
    return CustomPalette{palette.dark,    palette.backdrop, palette.base,
                         palette.surface, palette.overlay,  palette.subtext,
                         palette.text};
}

ThemeConfig::ThemeConfig() {
    const Palette palette = theme_palette(Theme::Midnight);
    custom = CustomPalette::from_palette(palette);
    gradient_start = palette.gradient[0];
    gradient_end = palette.gradient[1];
}

Palette ThemeConfig::palette() const {
    Palette resolved = theme_palette(theme);
    if (theme == Theme::Custom) {
        resolved.dark = custom.dark;
        resolved.backdrop = custom.backdrop;
        resolved.base = custom.base;
        resolved.surface = custom.surface;
        resolved.overlay = custom.overlay;
        resolved.subtext = custom.subtext;
        resolved.text = custom.text;
    }
    resolved.gradient = {gradient_start, gradient_end};
    return resolved;
}

void ThemeConfig::sync_gradient() {
    const Palette palette = theme_palette(theme);
    gradient_start = palette.gradient[0];
    gradient_end = palette.gradient[1];
}

void ThemeConfig::seed_custom(Theme from) { custom = CustomPalette::from_palette(theme_palette(from)); }

Color ThemeConfig::panel_fill(Color base) const {
    if (!gradient) {
        return base;
    }
    const float clamped = std::clamp(panel_opacity, 0.0f, 1.0f);
    return base.with_alpha(static_cast<std::uint8_t>(std::lround(clamped * 255.0f)));
}

Color ThemeConfig::secondary_text(const Palette& palette) const {
    return lerp(palette.subtext, palette.text, std::clamp(text_contrast, 0.0f, 1.0f));
}

Color heading_color(const Palette& palette, Color accent) {
    return lerp(palette.text, accent, 0.55f);
}

void apply(const ThemeConfig& config, Color accent) {
    const Palette palette = config.palette();
    const Color secondary = config.secondary_text(palette);
    const Color hovered = palette.overlay;
    const Color active = lerp(palette.overlay, accent, 0.45f);

    ImGuiStyle& style = ImGui::GetStyle();

    // one font at a base size, headings and hints push their own size on top of it
    style.FontSizeBase = body_size * std::clamp(config.text_scale, 0.7f, 2.0f);

    style.WindowPadding = ImVec2(10.0f, 10.0f);
    style.FramePadding = ImVec2(10.0f, 4.0f);
    style.ItemSpacing = ImVec2(8.0f, 6.0f);
    style.ItemInnerSpacing = ImVec2(6.0f, 4.0f);
    style.IndentSpacing = 18.0f;
    style.ScrollbarSize = 11.0f;
    style.GrabMinSize = 10.0f;

    style.WindowRounding = config.corner_radius;
    style.ChildRounding = config.corner_radius;
    style.FrameRounding = config.corner_radius;
    style.PopupRounding = config.corner_radius;
    style.GrabRounding = config.corner_radius;
    style.ScrollbarRounding = config.corner_radius;
    style.TabRounding = config.corner_radius;

    style.WindowBorderSize = 0.0f;
    style.ChildBorderSize = 0.0f;
    style.FrameBorderSize = 1.0f;
    style.PopupBorderSize = 1.0f;

    set(style, ImGuiCol_Text, palette.text);
    set(style, ImGuiCol_TextDisabled, secondary);
    set(style, ImGuiCol_WindowBg, config.panel_fill(palette.base));
    // children stay transparent so the window background, gradient included, shows through
    set(style, ImGuiCol_ChildBg, Color(0, 0, 0, 0));
    // popups and dropdowns stay opaque so their text keeps its contrast
    set(style, ImGuiCol_PopupBg, palette.base);
    set(style, ImGuiCol_Border, palette.overlay);
    set(style, ImGuiCol_BorderShadow, Color(0, 0, 0, 0));

    set(style, ImGuiCol_FrameBg, palette.surface);
    set(style, ImGuiCol_FrameBgHovered, hovered);
    set(style, ImGuiCol_FrameBgActive, active);

    set(style, ImGuiCol_TitleBg, palette.base);
    set(style, ImGuiCol_TitleBgActive, palette.surface);
    set(style, ImGuiCol_TitleBgCollapsed, palette.base);
    set(style, ImGuiCol_MenuBarBg, palette.surface);

    set(style, ImGuiCol_ScrollbarBg, Color(0, 0, 0, 0));
    set(style, ImGuiCol_ScrollbarGrab, palette.surface);
    set(style, ImGuiCol_ScrollbarGrabHovered, hovered);
    set(style, ImGuiCol_ScrollbarGrabActive, accent);

    set(style, ImGuiCol_CheckMark, accent);
    set(style, ImGuiCol_SliderGrab, accent);
    set(style, ImGuiCol_SliderGrabActive, lerp(accent, palette.text, 0.25f));

    set(style, ImGuiCol_Button, palette.surface);
    set(style, ImGuiCol_ButtonHovered, hovered);
    set(style, ImGuiCol_ButtonActive, active);

    set(style, ImGuiCol_Header, palette.surface);
    set(style, ImGuiCol_HeaderHovered, hovered);
    set(style, ImGuiCol_HeaderActive, active);

    set(style, ImGuiCol_Separator, palette.overlay);
    set(style, ImGuiCol_SeparatorHovered, lerp(palette.overlay, accent, 0.5f));
    set(style, ImGuiCol_SeparatorActive, accent);

    set(style, ImGuiCol_ResizeGrip, palette.surface);
    set(style, ImGuiCol_ResizeGripHovered, hovered);
    set(style, ImGuiCol_ResizeGripActive, accent);

    set(style, ImGuiCol_Tab, palette.surface);
    set(style, ImGuiCol_TabHovered, hovered);
    set(style, ImGuiCol_TabSelected, accent);

    set(style, ImGuiCol_TextLink, accent);
    set(style, ImGuiCol_TextSelectedBg, accent);
    set(style, ImGuiCol_InputTextCursor, palette.text);
    set(style, ImGuiCol_NavCursor, accent);

    // text drawn straight on the accent, so a bright accent does not get bright text
    set(style, ImGuiCol_DragDropTarget, contrasting_text(accent, palette));
}

void paint_gradient(const ThemeConfig& config, ImVec2 min, ImVec2 max) {
    if (!config.gradient || max.x <= min.x || max.y <= min.y) {
        return;
    }

    // corner colors are the gradient sampled at each corner's projection onto the gradient
    // axis, so the linear interpolation the gpu does between them reproduces the gradient
    constexpr float diagonal = 0.70710678f;
    ImVec2 axis;
    switch (config.gradient_direction) {
        case GradientDirection::Vertical: axis = ImVec2(0.0f, 1.0f); break;
        case GradientDirection::Horizontal: axis = ImVec2(1.0f, 0.0f); break;
        case GradientDirection::DiagonalDown: axis = ImVec2(diagonal, diagonal); break;
        case GradientDirection::DiagonalUp: axis = ImVec2(diagonal, -diagonal); break;
    }

    const std::array<ImVec2, 4> corners{
        ImVec2(min.x, min.y),
        ImVec2(max.x, min.y),
        ImVec2(max.x, max.y),
        ImVec2(min.x, max.y),
    };

    std::array<float, 4> projections{};
    for (std::size_t i = 0; i < corners.size(); ++i) {
        projections[i] = corners[i].x * axis.x + corners[i].y * axis.y;
    }
    const float lowest = *std::ranges::min_element(projections);
    const float highest = *std::ranges::max_element(projections);
    const float span = highest - lowest;

    std::array<ImU32, 4> colors{};
    for (std::size_t i = 0; i < colors.size(); ++i) {
        const float t = span > 1e-6f ? (projections[i] - lowest) / span : 0.0f;
        colors[i] = lerp(config.gradient_start, config.gradient_end, t).u32();
    }

    ImGui::GetBackgroundDrawList()->AddRectFilledMultiColor(min, max, colors[0], colors[1],
                                                            colors[2], colors[3]);
}

}  // namespace dl::ui
